File: app/src/main/java/ua/hospes/rtm/data/race/mapper/RaceMapper.java
package ua.hospes.rtm.data.race.mapper;

import android.database.Cursor;

import ua.hospes.rtm.core.db.tables.Race;
import ua.hospes.rtm.data.race.models.RaceItemDb;
import ua.hospes.rtm.domain.race.models.RaceItem;
import ua.hospes.rtm.domain.sessions.Session;
import ua.hospes.rtm.domain.team.Team;
import ua.hospes.rtm.utils.Optional;

public class RaceMapper {
    public static RaceItemDb map(Cursor cursor) {
        int teamId = cursor.getInt(cursor.getColumnIndex(Race.TEAM_ID.name()));
        RaceItemDb result = new RaceItemDb(teamId);
        result.setId(cursor.getInt(cursor.getColumnIndex(Race.ID.name())));
        result.setTeamNumber(cursor.getInt(cursor.getColumnIndex(Race.TEAM_NUMBER.name())));
        result.setSessionId(cursor.getInt(cursor.getColumnIndex(Race.SESSION_ID.name())));
        result.setOrder(cursor.getInt(cursor.getColumnIndex(Race.ORDER.name())));
        return result;
    }

    public static RaceItem map(RaceItemDb db, Optional<Team> team, Optional<Session> session) {
        RaceItem result = new RaceItem(db.getId(), team.get());
        result.setTeamNumber(db.getTeamNumber());
        result.setSession(session.get());
        return result;
    }

    public static RaceItemDb map(RaceItem item) {
        RaceItemDb db = new RaceItemDb(item.getTeam().getId());

        db.setId(item.getId());
        db.setTeamNumber(item.getTeamNumber());
        if (item.getSession() != null)
            db.setSessionId(item.getSession().getId());

        return db;
    }
}

File: app/src/main/java/ua/hospes/rtm/db/race/RaceMapper.kt
package ua.hospes.rtm.db.race

import android.database.Cursor

import ua.hospes.rtm.db.tables.Race
import ua.hospes.rtm.domain.race.models.RaceItem
import ua.hospes.rtm.domain.sessions.Session
import ua.hospes.rtm.domain.team.Team
import ua.hospes.rtm.utils.Optional

object RaceMapper {
    fun map(cursor: Cursor): RaceItemDb {
        val teamId = cursor.getInt(cursor.getColumnIndex(Race.TEAM_ID.name()))
        val result = RaceItemDb(teamId)
        result.id = cursor.getInt(cursor.getColumnIndex(Race.ID.name()))
        result.teamNumber = cursor.getInt(cursor.getColumnIndex(Race.TEAM_NUMBER.name()))
        result.sessionId = cursor.getInt(cursor.getColumnIndex(Race.SESSION_ID.name()))
        result.order = cursor.getInt(cursor.getColumnIndex(Race.ORDER.name()))
        return result
    }

    fun map(db: RaceItemDb, team: Optional<Team>, session: Optional<Session>): RaceItem {
        val result = RaceItem(db.id, team.get()!!)
        result.teamNumber = db.teamNumber
        result.session = session.get()
        return result
    }

    fun map(item: RaceItem): RaceItemDb {
        val db = RaceItemDb(item.team.id!!)

        db.id = item.id
        db.teamNumber = item.teamNumber
        if (item.session != null)
            db.sessionId = item.session?.id!!

        return db
    }
}